### PR TITLE
Disable throttling for non-performance passes

### PR DESCRIPTION
**Pull request by paulirish**
This is the approach that I ended up with after playing around a bit.

* `useThrottling` prop in the config for each pass
* if that is true and the disable CLI flags aren't set, then we'll turn it on.
* if the prop is false for a pass we'll definitely make sure throttling is off
* when `goOnline` comes back it'll restore whatever throttling should be for that pass.
* `setThrottling` method in driver does the work. the old `beginEmulation` now defers to it for some parts.

wdyt?

### CODE REVIEW

File: lighthouse-core/gather/driver.js [hunk 1]
```diff
@@ -673,21 +673,31 @@ class Driver {
     return this.sendCommand('Runtime.enable');
   }
 
-  beginEmulation(flags) {
+  beginEmulation(options) {
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver.js]: any reason to switch from flags to options? looks like this just uses flags

File: lighthouse-core/gather/driver.js [hunk 2]
```diff
@@ -673,21 +673,31 @@ class Driver {
...
+      return this.setThrottling(options.flags, {useThrottling: true});
+    });
+  }
+
+  setThrottling(cliFlags, passConfig) {
     const emulations = [];
 
-    if (!flags.disableDeviceEmulation) {
-      emulations.push(emulation.enableNexus5X(this));
+    if (!passConfig.useThrottling) {
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver.js]: this is one case I think having it nested would make it more readable, took me a bit to spot the early return (though that may be more diff noise and it's easier without the diff)

thoughts on the following?
```
if (useThrottling) {
  if (!disableNetwork) emulations.push()
  if (!disableCpu) emulations.push
} else {
  emulations.push(...)
}

return Promise.all()
```

File: lighthouse-core/gather/driver.js [hunk 3]
```diff
@@ -696,11 +706,11 @@ class Driver {
...
   goOffline() {
-    return this.sendCommand('Network.enable').then(_ => {
-      return emulation.goOffline(this);
-    }).then(_ => {
-      this.online = false;
-    });
+    return this.sendCommand('Network.enable')
+      .then(_ => emulation.goOffline(this))
+      .then(_ => {
+        this.online = false;
```
Comment (patrickhulce) [inline lighthouse-core/gather/driver.js]: nit: could probably one-line this too

File: lighthouse-core/gather/driver.js [hunk 4]
```diff
@@ -673,21 +673,31 @@ class Driver {
...
+    return Promise.resolve().then(_ => {
+      if (!options.flags.disableDeviceEmulation) {
+        return emulation.enableNexus5X(this);
+      }
+    }).then(_ => {
+      return this.setThrottling(options.flags, {useThrottling: true});
+    });
+  }
+
+  setThrottling(cliFlags, passConfig) {
```
Comment (ebidel) [inline lighthouse-core/gather/driver.js]: should passConfig just be a boolean, `useThrottling`?

Comment (ebidel) [inline lighthouse-core/gather/driver.js]: Ah, see you're passing in the `options.config` below. Guess it's fine.

File: lighthouse-core/gather/driver.js (hunk 2 above)
Comment (paulirish) [inline lighthouse-core/gather/driver.js]: so much better. done.

**Pull request merged**